src/main/java/io/github/thebusybiscuit/slimefun4/api/events/PreCanUnlockResearchEvent.java

```diff
@@ -0,0 +1,82 @@
+package io.github.thebusybiscuit.slimefun4.api.events;
+
+import javax.annotation.Nonnull;
+import javax.annotation.ParametersAreNonnullByDefault;
+import org.bukkit.entity.Player;
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+import org.apache.commons.lang.Validate;
+import me.mrCookieSlime.Slimefun.Objects.SlimefunItem.SlimefunItem;
+import io.github.thebusybiscuit.slimefun4.core.researching.Research;
+import io.github.thebusybiscuit.slimefun4.implementation.guide.BookSlimefunGuide;
+import io.github.thebusybiscuit.slimefun4.implementation.guide.ChestSlimefunGuide;
+
+/**
+ * This {@link Event} is called whenever a {@link Player} clicks to unlock a research.
+ * This is called before {@link Research#canUnlock(Player)}.
+ * The event is not called for the cheat sheet.
+ *
+ * @author uiytt
+ *
+ * @see ChestSlimefunGuide
+ * @see BookSlimefunGuide
+ *
+ */
+public class PreCanUnlockResearchEvent extends Event implements Cancellable {
```

Comment:
This class name doesn't really follow any conventions here...
Perhaps just  `PlayerPreResearchEvent`?

Reply:
Thanks, now I'm terrible at naming things